[leeyi45] iP

leeyi45 opened the pull request:
# Duke
A text based chatbot for all your task handling needs!

Duke can handle the following types of tasks
- Todos: Tasks with no deadlines
- Deadlines: Tasks with deadlines!
- Events: Tasks with a start and end time

Using Duke is just as easy as:
1. Clone [this repository](github.com/leeyi45/iP)!
1. Compiling the code with `javac`!
1. Running Duke with `java -jar Duke.jar`!

*Simple as that!*

>It really is that simple! 😱 
-some random user

## Some dev pointers
To add a new task type, remember to tag it with the TaskAnno annotation!
```java
@TaskAnno("M") // Don't forget this!
public class NewTask {
  // implementation ....
}
```
You can refer to the Java page on annotations for more information: https://docs.oracle.com/javase/tutorial/java/annotations/
## Tasks left to do
- [X] ~Added ability to save tasks~
- [ ] Add a GUI to Duke

Commit:
Add Gradle support

leeyi45 committed:
Level-1

leeyi45 committed:
Level-2

leeyi45 committed:
Level-3

leeyi45 committed:
Level 4

leeyi45 committed:
Level 5

leeyi45 committed:
Added more error handling

leeyi45 committed:
Level 6

leeyi45 committed:
Fixing some bugs

leeyi45 committed:
Ignoring vscode settings

leeyi45 committed:
Adding documentation

leeyi45 committed:
Text testing

leeyi45 committed:
Refactored code to be more modular

leeyi45 committed:
Finished transferring all commands to new foramt

leeyi45 committed:
Fixed bug where each command could only run once

leeyi45 committed:
Revert "Refactored code to be more modular"

This reverts commit 3090dafd303b721c3ff9f7b65bd9f05d2d1132fc.

leeyi45 committed:
Cleaned master branch

leeyi45 committed:
Level-7

leeyi45 committed:
Let Duke understand times

Parse date and time information as actual TemporalAccessor objects

leeyi45 committed:
Make Assumptions about dates

If a time field or date field was not provided by the user, fill the missing field with LocalDateTime.now()

leeyi45 committed:
Validate DateTime inputs

Make sure that deadlines can't be in the past, and that events cannot end before they start

leeyi45 committed:
Merge branch 'branch-Level-8'

leeyi45 committed:
Merge branch 'after-level-8'

leeyi45 committed:
A-Packages

Refactored the code to be more modular

leeyi45 committed:
A-MoreOOP

I love OOP so much that I refactored the entire thing into more classes!

leeyi45 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

leeyi45 committed:
Added gradle support

leeyi45 committed:
Added tests

leeyi45 committed:
Fixes gradle properties

The JDK path for gradle wasn't configured correctly before. This commit fixes that

leeyi45 committed:
Add checkstyle

This commit adds checkstyle checks to the project.

leeyi45 committed:
Add JavaDocs

Added documentation for most of the code, which was missing previously.

leeyi45 committed:
Refactors index only commands

Extract out all commands that only need 1 index parameter to ensure that their functionality is unified

leeyi45 committed:
Add FindCommand

leeyi45 committed:
Merge branch 'branch-A-CodingStandard'

leeyi45 committed:
Fix broken tests

Certain tests were failing because the test themselves were broken. Now they should be working with no issues

leeyi45 committed:
Merge branch 'branch-Level-9'

leeyi45 committed:
Finish merge with level 9

leeyi45 committed:
Update Duke.java

Fix incorrect imports

leeyi45 committed:
Minor aesthetic changes

Edits code to follow checkstyle guidelines, as well as changes to the output of certain console statements

leeyi45 committed:
Remove /to after /from restriction

The event command previously required that /to came after /from. This commit removes that restriction

szejiancheng commented on src/main/java/duke/Program.java:
```diff
@@ -0,0 +1,34 @@
+            } catch (IOException e) {
+                System.out.format("An IOException occurred while reading your tasks from disk: '%s'\n", e.getMessage());
+            } catch (ClassNotFoundException e) {
+                System.out.println("Your data file was corrupted, so we were unable to read your tasks from there");
```
will the program exit when an error is encountered? How about when the user is using this program for the first time? Will data.dat already be initialized?

szejiancheng commented on src/main/java/duke/Program.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.List;
+
+import duke.task.Task;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/Utils.java:
```diff
@@ -0,0 +1,121 @@
+        Function<String, Optional<LocalTime>> timeParser = createParser(timeFormats, LocalTime::parse);
+
+        if (str0 == null && str1 == null) {
+            throw new IllegalArgumentException("Both str0 and str1 cannot be null!", null);
```
Will this exception be thrown to the user? Where else could this exception be caught and handled with?

szejiancheng commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public abstract class Command implements BiConsumer<String[], Duke> {
+    /**
+     * Represents errors in the user's input. Throw this exception within {@link #execute() execute}
```
Love the javadocs links!

szejiancheng commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,39 @@
+import duke.Utils;
+import duke.task.Task;
+import duke.task.TaskList;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/commands/indexedCommand/IndexedCommand.java:
```diff
@@ -0,0 +1,30 @@
+import duke.Duke;
+import duke.commands.Command;
+import duke.task.Task;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/commands/indexedCommand/MarkCommand.java:
```diff
@@ -0,0 +1,16 @@
+
+import duke.Duke;
+import duke.task.Task;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/commands/indexedCommand/UnmarkCommand.java:
```diff
@@ -0,0 +1,16 @@
+
+import duke.Duke;
+import duke.task.Task;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/commands/taskCommand/DeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+import duke.Duke;
+import duke.Utils;
+import duke.task.Deadline;
+
```
May want to include JavaDocs here!

szejiancheng commented on src/main/java/duke/commands/taskCommand/DeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+                args[index + 1],
+                index + 2 >= args.length ? null : args[index + 2]
+            );
+            validate(deadline.isAfter(LocalDateTime.now()), "I can't create something's that's due in the past!");
```
nice one checking for this, inspired me to do something similar!

szejiancheng reviewed (commented):
Cool implementation of Duke! Some concerns with exception catching but overall looks great!

leeyi45 committed:
Added a UI to Duke

leeyi45 committed:
Fixes AddTaskWindow

leeyi45 committed:
Fix tasks not saving and loading correctly

leeyi45 committed:
Merge branch 'branch-A-Varargs'

leeyi45 committed:
Add assertions

leeyi45 committed:
Update App.java

Using the quit button did not properly trigger the exit code. It should now.

leeyi45 committed:
Rename execute method in Command

Originally to execute a command, the accept() method is called. This is not reflective of the nature of executing the command.

Now the command type exposes an execute function.

leeyi45 committed:
Abstract printing methods

Commands would print output directly to System.out. This commit allows commands to have customized output functions that can redirect output to other places.

leeyi45 committed:
Merge branch 'master' into branch-C-Priority

leeyi45 committed:
Add Priority Option

Added a new parser system that is able to extract each option and its corresponding arguments, making the parsing of commands easier.

leeyi45 committed:
Change package names

Previously, the names indexedCommand and taskCommand were in use. This commit will change them to fully lowercase to comply with the coding standard.

leeyi45 committed:
Update README.md

leeyi45 committed:
Add files via upload

leeyi45 committed:
Revamp the UI

leeyi45 committed:
Merge branch 'master' of https://github.com/leeyi45/ip

leeyi45 committed:
Update README.md

leeyi45 committed:
Update README.md
